torch/csrc/dynamo/guards.cpp

```diff
@@ -574,6 +574,371 @@ static struct PyModuleDef _module = {
     -1,
     _methods};
 
+/**
+ * Stores relevant guard debug information, e.g., failure str for a LeafGuard
+ * failure. The data structure is also accessible in Python.
+ */
+struct GuardDebugInfo {
+  GuardDebugInfo(int num_guards_executed, std::string failure_reason) {
+    this->num_guards_executed = num_guards_executed;
+    this->failure_reason = failure_reason;
+  }
+
+  std::string repr() {
+    return "GuardDebugInfo(num_guards_executed=" +
+        std::to_string(num_guards_executed) +
+        ", failure_reason=" + failure_reason + ")";
+  }
+
+  // Failure reason for a leaf guard.
+  std::string failure_reason;
+
+  // Total number of executed guards so far.
+  int num_guards_executed;
+};
+
+/**
+ * Base class for the terminal or leaf guard in the GuardManager hierarchy.
+ */
+class LeafGuard {
+ public:
+  // Runs the guard and prepares a GuardDebugInfo object.
+  std::pair<bool, GuardDebugInfo> check_with_debug_info(py::object value) {
+    bool result = run_guards(value);
+    if (result == false) {
+      std::string reason = get_failure_reason(value);
+      return std::make_pair(result, GuardDebugInfo(0, reason));
+    }
+    return std::make_pair(result, GuardDebugInfo(0, "PASS"));
+  }
+
+  // Runs the guard
+  bool check(py::object value) {
+    return run_guards(value);
+  }
+
+  virtual bool run_guards(py::object value) = 0;
+  virtual std::string get_failure_reason(py::object value) = 0;
+  virtual ~LeafGuard() = default;
+};
+
+/**
+ * Represents a leaf guard that accepts the python guard check function. We
+ * would like to have most of the guards in C++ (to avoid a Python function
+ * call).  But, it will take some time to reach that goal. Also, there might be
+ * cases where its too tedious to write an equivalent C++ guard.
+ *
+ * PythonLambdaGuard allows us to gradually move to C++. We can start from all
+ * guards of type PythonLambaGuard and incrementally move expensive guards to
+ * C++.
+ */
+class PythonLambdaGuard : public LeafGuard {
+ public:
+  PythonLambdaGuard(py::object guard_check_fn, py::object print_failure_fn) {
+    if (py::isinstance<py::function>(guard_check_fn) &&
+        py::isinstance<py::function>(print_failure_fn)) {
+      _guard_check_fn = py::cast<py::function>(guard_check_fn);
+      _print_failure_fn = py::cast<py::function>(print_failure_fn);
+    } else {
+      throw py::type_error("PythonLambdaGuard expects callables");
+    }
+  }
+
+  // Runs the lambda function with the current f_locals value.
+  bool run_guards(py::object value) override {
+    return py::cast<bool>(_guard_check_fn(value));
+  }
+
+  std::string get_failure_reason(py::object value) override {
+    return py::cast<std::string>(_print_failure_fn(value));
+  }
```

Comment:
string_view, here and everywhere

Reply:
So I tried `string_view` and got some issues accessing in Python using pybind - https://zpz.github.io/blog/python-cpp-pybind11-stringview/

Though I can work harder to fix the `string_view` + pybind issue, but I don't think we need to complicate it. The `std::string` part of the codebase is not in the hot path. We need string only for recompilation reasoning which we will soon move to convert_frame.py - https://github.com/pytorch/pytorch/issues/108950

So I am going to not implement string_view here.